text/0000-constants-in-patterns.md

```diff
@@ -0,0 +1,306 @@
+- Do something that violates the core design principles laid out in the "Motivation" section.
+  This is considered a no-go by this RFC:
+  having possible behavior changes upon "outlining" a fieldless enum variant into a constant is too big of a footgun,
+  and allowing matching without opt-in from the crate that defines the type makes abstraction and semver too tricky to maintain (consider that private fields would be compared when matching).
+
+# Prior art
+[prior-art]: #prior-art
+
+[RFC 1445][rfc-1445] defines basically the same checks as this RFC; this RFC merely spells them out more clearly for cases the old RFC did not explicitly over (nested user-defined types, raw pointers),
+and adjusts to decisions that have been made in the mean time (accepting floating-point patterns).
+
+This RFC came out of discussions in a [t-lang design meeting](https://github.com/rust-lang/lang-team/issues/220).
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+None at this time.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+- At some point, we might want to stabilize the `StructuralPartialEq` trait.
+  There are however plenty of open questions here:
+  + Should `StructuralPartialEq` be an unsafe trait?
+    That trait has a clear semantic meaning, so making it `unsafe` to be able to rely on it for soundness is appealing.
+    In particular, we could then be *sure* that pattern matching always has the same semantics as `==`.
+    With the trait being safe, it is actually possible to write patterns that behave different from the `PartialEq` that is explicitly defined and intentionally exposed on that type,
+    but only if one of the involved crates implements `StructuralPartialEq` incorrectly.
+    This can lead to semver issues and logic bugs, but that is all allowed for safe traits.
+    However, this also means unsafe code cannot rely on `==` and pattern matching having the same behavior.
+    To make the trait unsafe, the logic for `derive(PartialEq)` should use `#[allow_internal_unsafe]` to still pass `forbid(unsafe_code)`.
+  + What should the `StructuralPartialEq` trait be called?
+    The current name can be considered confusing because the trait reflects a shallow property, while the value-based check performed when a constant is used in a pattern is defined recursively.
+  + Trait bounds `T: StructuralPartialEq` seem like a strange concept. Should we really allow them?
+
+- To avoid interpreting `derive(PartialEq)` as a semver-stable promise of being able to structurally match this type,
+  we could introduce an explicit `derive(PartialEq, StructuralPartialEq)`.
+  However, that would be a massive breaking change, so it can only be done over an edition boundary.
+  It probably would also want to come with some way to say "derive me all the usual traits" so that one does not have to spell out so many trait names.
```

Comment:
One potential alternative to this is to allow negative impls for `StructuralPartialEq` for derived `PartialEq`, and/or perhaps to allow typing `derive(PartialEq, !StructuralPartialEq)`.

If we think of `StructuralPartialEq` "like an auto trait" since it's infectious, this should make some amount of sense.